_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.web.server.ServerHttpSecurity;
+import org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository;
+import org.springframework.security.oauth2.client.web.server.DefaultServerOAuth2AuthorizationRequestResolver;
+import org.springframework.security.oauth2.client.web.server.ServerOAuth2AuthorizationRequestResolver;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest;
+import org.springframework.security.web.server.SecurityWebFilterChain;
+
+import java.util.function.Consumer;
+
+@Configuration
+public class SecurityConfiguration {
+
+    @Value("${okta.oauth2.audience:}")
+    private String audience;
+
+    private final ReactiveClientRegistrationRepository clientRegistrationRepository;
+
+    public SecurityConfiguration(ReactiveClientRegistrationRepository clientRegistrationRepository) {
+        this.clientRegistrationRepository = clientRegistrationRepository;
+    }
+
+    @Bean
+    public SecurityWebFilterChain filterChain(ServerHttpSecurity http) throws Exception {
+        http
+            .authorizeExchange(authz -> authz
+                .anyExchange().authenticated()
+            )
+            .oauth2Login(oauth2 -> oauth2
+                .authorizationRequestResolver(
+                    authorizationRequestResolver(this.clientRegistrationRepository)
+                )
+            );
+        return http.build();
+    }
+
+    private ServerOAuth2AuthorizationRequestResolver authorizationRequestResolver(
+        ReactiveClientRegistrationRepository clientRegistrationRepository) {
+
+        var authorizationRequestResolver =
+            new DefaultServerOAuth2AuthorizationRequestResolver(clientRegistrationRepository);
+        authorizationRequestResolver.setAuthorizationRequestCustomizer(
+            authorizationRequestCustomizer());
+
+        return authorizationRequestResolver;
+    }
+
+    private Consumer<OAuth2AuthorizationRequest.Builder> authorizationRequestCustomizer() {
+        return customizer -> customizer
+            .additionalParameters(params -> params.put("audience", audience));
+    }
+}
+```
+
+I was able to get the OpenFeign client to work by adding a couple of properties to enable OAuth:
+
+```properties
+spring.cloud.openfeign.oauth2.enabled=true
+spring.cloud.openfeign.oauth2.clientRegistrationId=okta
+```
+
+To make Spring Cloud Gateway pass the access token downstream, I added `TokenRelay` to its default filters in `application.yml`:
+
+```yaml
+spring:
+  cloud:
+    gateway:
+      discovery:
+        locator:
+          enabled: true
+      default-filters:
+        - TokenRelay
+      routes: ...
+```
+
+I added a `WebClientConfiguration` class to configure `WebClient` to include the access token with its requests.
```

Comment:
```suggestion
I updated the `WebClientConfiguration` class to configure `WebClient` to include the access token with its requests.
```
since this is required earlier on